mindmeld/components/entity_resolver.py

```diff
@@ -42,60 +50,277 @@
+                    msg = "Synonym %s specified in %s entity map multiple times"
+                    logger.debug(msg, cname, entity_type)
+                cnames_for_syn = syn_map.get(norm_alias, [])
+                cnames_for_syn.append(cname)
+                syn_map[norm_alias] = list(set(cnames_for_syn))
+
+        # extend synonyms map by adding keys which are lowercases or titlecase of the existing keys
+        if _form_explorer:
+            msg = "Adding additional form of the whitelist and cnames to list of possible synonyms"
+            logger.info(msg)
+            initial_num_syns = len(syn_map)
+            aug_syn_map = {}
+            for alias, alias_map in syn_map.items():
+                other_aliases = _form_explorer(alias)
+                for other_alias in other_aliases:
+                    if other_alias not in syn_map:
+                        aug_syn_map.update({other_alias: alias_map})
+            syn_map.update(aug_syn_map)
+            final_num_syns = len(syn_map)
+            msg = "Added %d additional synonyms by lower-casing. Synonyms increased from %d to %d"
+            logger.info(msg, final_num_syns - initial_num_syns, initial_num_syns, final_num_syns)
+
+        return {"items": item_map, "synonyms": syn_map}
+
+    @abstractmethod
+    def _fit(self, clean):
+        raise NotImplementedError
+
+    def fit(self, clean=False):
+        """Fits the resolver model, if required
+
+        Args:
+            clean (bool, optional): If ``True``, deletes and recreates the index from scratch
+                                    with synonyms in the mapping.json.
+        """
+
+        if self.ready:
+            return
+
+        if self._is_system_entity or not self.entity_map.get("entities", []):
+            return
+
+        self._fit(clean)
+        self.ready = True
+
+    @abstractmethod
+    def _predict(self, nbest_entities, top_n):
+        raise NotImplementedError
+
+    def predict(self, entity, top_n: int = 20):
+        """Predicts the resolved value(s) for the given entity using the loaded entity map or the
+        trained entity resolution model.
+
+        Args:
+            entity (Entity, tuple): An entity found in an input query, or a list of n-best entity \
+                objects.
+            top_n (int): maximum number of results to populate
+
+        Returns:
+            (list): The top n resolved values for the provided entity.
+        """
+        if isinstance(entity, (list, tuple)):
+            top_entity = entity[0]
+            nbest_entities = tuple(entity)
+        else:
+            top_entity = entity
+            nbest_entities = tuple([entity])
+
+        if self._is_system_entity:
+            # system entities are already resolved
+            return [top_entity.value]
+
+        if not self.entity_map.get("entities", []):
+            return []
+
+        results = self._predict(nbest_entities, top_n)
+
+        if results:
```

Comment:
can results be `None`? If not, we could just do this:
```
results = results[:top_n]
if len(results) < top_n:
                logger.info(
                    "Retrieved only %d entity resolutions instead of asked number %d for "
                    "entity %r of type %r",
                    len(results), top_n, nbest_entities[0].text, self.type,
                )

```

Reply:
else, we can do this:
```
if not results: return None
results = results[:top_n]
if len(results) < top_n:
                logger.info(
                    "Retrieved only %d entity resolutions instead of asked number %d for "
                    "entity %r of type %r",
                    len(results), top_n, nbest_entities[0].text, self.type,
                )


```

Reply:
From my understanding (and some experiments), Elastic Search fails to resolve some entities, in which case None is returned. Also, that's a case with Exact match algo. So the code snippet is updated as follows

```
        results = self._predict(nbest_entities, top_n)

        if not results:
            return None

        results = results[:top_n]
        if len(results) < top_n:
            logger.info(
                "Retrieved only %d entity resolutions instead of asked number %d for "
                "entity %r of type %r",
                len(results), top_n, nbest_entities[0].text, self.type,
            )
        
        return results
```